### PR TITLE
Core: add `BinaryInteger` protocol

### PR DESCRIPTION
**Pull request by MaxDesiatov**

### CODE REVIEW

**Comment by compnerd** [comment]
Let's split out `Strideable` out of this addition.  I definitely would like to get a declaration of `BinaryInteger` added, this will allow us to move the definitions of `&` and `&=` out of the integer types and into the protocol conformance which introduces it.

**Comment by MaxDesiatov** [comment]
Done. Added a merge commit for quick conflict resolution, but it should be gone when you squash and merge.

File: Sources/Core/BinaryInteger.swift [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  static func /= (_ lhs: inout Self, _ rhs: Self)
+  static func % (_ lhs: Self, _ rhs: Self) -> Self
+  static func %= (_ lhs: inout Self, _ rhs: Self)
+
+  override static func + (_ lhs: Self, _ rhs: Self) -> Self
+  override static func += (_ lhs: inout Self, _ rhs: Self)
+  override static func - (_ lhs: Self, _ rhs: Self) -> Self
+  override static func -= (_ lhs: inout Self, _ rhs: Self)
+  override static func * (_ lhs: Self, _ rhs: Self) -> Self
+  override static func *= (_ lhs: inout Self, _ rhs: Self)
```
Comment (compnerd) [inline Sources/Core/BinaryInteger.swift:27]: Let's hoist the `override` functions.

File: Sources/Core/BinaryInteger.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  static var isSigned: Bool { get }
+
+  init<T: BinaryInteger>(_ source: T)
+
+  init<T: BinaryInteger>(truncatingIfNeeded source: T)
+
+  init<T: BinaryInteger>(clamping source: T)
+
+  var bitWidth: Int { get }
+  var trailingZeroBitCount: Int { get }
```
Comment (compnerd) [inline Sources/Core/BinaryInteger.swift:15]: Put the ivars before the methods (constructors or functions).

File: Sources/Core/BinaryInteger.swift [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  override static func * (_ lhs: Self, _ rhs: Self) -> Self
+  override static func *= (_ lhs: inout Self, _ rhs: Self)
+
+  static prefix func ~ (_ x: Self) -> Self
+  static func & (_ lhs: Self, _ rhs: Self) -> Self
+  static func &= (_ lhs: inout Self, _ rhs: Self)
+  static func | (_ lhs: Self, _ rhs: Self) -> Self
+  static func |= (_ lhs: inout Self, _ rhs: Self)
+  static func ^ (_ lhs: Self, _ rhs: Self) -> Self
+  static func ^= (_ lhs: inout Self, _ rhs: Self)
```
Comment (compnerd) [inline Sources/Core/BinaryInteger.swift:35]: Mind creating a separate group below this for the mutate and assign variants?

File: Sources/Core/BinaryInteger.swift [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  static func >> <RHS: BinaryInteger>(_ lhs: Self, _ rhs: RHS) -> Self
+  static func >>= <RHS: BinaryInteger>(_ lhs: inout Self, _ rhs: RHS)
+  static func << <RHS: BinaryInteger>(_ lhs: Self, _ rhs: RHS) -> Self
+  static func <<=< RHS: BinaryInteger > (_ lhs: inout Self, _ rhs: RHS)
```
Comment (compnerd) [inline Sources/Core/BinaryInteger.swift:40]: Similar

File: Sources/Core/BinaryInteger.swift [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+  func quotientAndRemainder(dividingBy rhs: Self)
+    -> (quotient: Self, remainder: Self)
```
Comment (compnerd) [inline Sources/Core/BinaryInteger.swift:43]:
```suggestion
      -> (quotient: Self, remainder: Self)
```

File: Sources/Core/Operators.swift [hunk 6]
```diff
@@ -74,11 +76,28 @@ infix operator ||: LogicalDisjunctionPrecedence
...
 infix operator -=: AssignmentPrecedence
 infix operator *=: AssignmentPrecedence
+infix operator /=: AssignmentPrecedence
+infix operator %=: AssignmentPrecedence
+infix operator |=: AssignmentPrecedence
+infix operator ^=: AssignmentPrecedence
+infix operator <<=: AssignmentPrecedence
+infix operator &<<=: AssignmentPrecedence
+infix operator >>=: AssignmentPrecedence
+infix operator &>>=: AssignmentPrecedence
```
Comment (compnerd) [inline Sources/Core/Operators.swift:98]: What do you think of splitting up the bit-shift and assign and ignoring-overflow variants into separate groups?

File: Sources/Core/Operators.swift [hunk 7]
```diff
@@ -74,11 +76,28 @@ infix operator ||: LogicalDisjunctionPrecedence
...
+
+infix operator <<: BitwiseShiftPrecedence
+infix operator &<<: BitwiseShiftPrecedence
+infix operator >>: BitwiseShiftPrecedence
+infix operator &>>: BitwiseShiftPrecedence
```
Comment (compnerd) [inline Sources/Core/Operators.swift:103]: Similar

File: Sources/Core/BinaryInteger.swift [hunk 8]
```diff
@@ -0,0 +1,46 @@
+// Copyright © 2022 Max Desiatov <max@desiatov.com>.
+// All Rights Reserved.
```
Comment (compnerd) [inline Sources/Core/BinaryInteger.swift:2]:
```suggestion
// Copyright © 2022 μSwift Authors.  All Rights Reserved.
```

**Review by compnerd — APPROVED** [review]
Thanks!

**Pull request merged**